src/communication/websocket/session.cpp

```diff
@@ -19,42 +19,78 @@
 #include <spdlog/spdlog.h>
 #include <boost/asio/bind_executor.hpp>
 #include <boost/beast/core/buffers_to_string.hpp>
+#include <boost/beast/core/stream_traits.hpp>
 #include <json/json.hpp>
 
 #include "communication/context.hpp"
 #include "utils/logging.hpp"
 
 namespace communication::websocket {
 namespace {
-void LogError(boost::beast::error_code ec, const std::string_view what) {
+void LogError(const boost::beast::error_code ec, const std::string_view what) {
   spdlog::warn("Websocket session failed on {}: {}", what, ec.message());
 }
 }  // namespace
 
-void Session::Run() {
-  ws_.set_option(boost::beast::websocket::stream_base::timeout::suggested(boost::beast::role_type::server));
+std::variant<Session::PlainWebSocket, Session::SSLWebSocket> Session::CreateWebSocket(tcp::socket &&socket,
+                                                                                      ServerContext &context) {
+  if (context.use_ssl()) {
+    ssl_context_.emplace(context.context_clone());
+    return Session::SSLWebSocket{std::move(socket), *ssl_context_};
+  }
 
-  ws_.set_option(boost::beast::websocket::stream_base::decorator(
-      [](boost::beast::websocket::response_type &res) { res.set(boost::beast::http::field::server, "Memgraph WS"); }));
+  return Session::PlainWebSocket{std::move(socket)};
+}
 
-  // Accept the websocket handshake
-  boost::beast::error_code ec;
-  ws_.accept(ec);
-  if (ec) {
-    return LogError(ec, "accept");
+Session::Session(tcp::socket &&socket, ServerContext &context, SafeAuth auth)
+    : ws_(CreateWebSocket(std::move(socket), context)), strand_{boost::asio::make_strand(GetExecutor())}, auth_{auth} {}
+
+bool Session::Run() {
+  ExecuteForWebsocket([](auto &&ws) {
+    ws.set_option(boost::beast::websocket::stream_base::timeout::suggested(boost::beast::role_type::server));
+
+    ws.set_option(boost::beast::websocket::stream_base::decorator([](boost::beast::websocket::response_type &res) {
+      res.set(boost::beast::http::field::server, "Memgraph WS");
+    }));
+  });
+
+  if (auto *ssl_ws = std::get_if<SSLWebSocket>(&ws_)) {
```

Comment:
I think with an explicit condition it would be better:

```suggestion
  if (auto *ssl_ws = std::get_if<SSLWebSocket>(&ws_); ssl_ws != nullptr) {
```

or at least
```suggestion
  if (auto *ssl_ws = std::get_if<SSLWebSocket>(&ws_); ssl_ws) {
```
I know that the result is the same, but these two make it more readable in my opinion (I like explicit more than implicit, so I might be alone.)